File: app/src/main/java/im/point/dotty/repository/RecentRepo.java
package im.point.dotty.repository;

import java.util.List;

import im.point.dotty.db.RecentPostDao;
import im.point.dotty.mapper.Mapper;
import im.point.dotty.model.RecentPost;
import im.point.dotty.network.MetaPost;
import im.point.dotty.network.ObservableCallBackAdapter;
import im.point.dotty.network.PointAPI;
import im.point.dotty.network.PostsReply;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

class RecentRepo implements Repository<RecentPost> {
    private final PointAPI api;
    private final RecentPostDao recentPostDao;
    private final String token;
    private final Mapper<RecentPost, MetaPost> mapper;

    RecentRepo(PointAPI api, String token, RecentPostDao recentPostDao, Mapper<RecentPost, MetaPost> mapper) {
        this.api = api;
        this.token = token;
        this.recentPostDao = recentPostDao;
        this.mapper = mapper;
    }

    @Override
    public Flowable<List<RecentPost>> getAll() {
        return recentPostDao.getAll();
    }

    @Override
    public Single<List<RecentPost>> fetch() {
        Observable<PostsReply> source = Observable.create(emitter -> {
            api.getRecent(token, "").enqueue(new ObservableCallBackAdapter<>(emitter));
        });
        return source.observeOn(Schedulers.io())
                .flatMap(postsReply -> Observable.fromIterable(postsReply.getPosts()))
                .map(mapper::map)
                .toList()
                .doOnSuccess(recentPosts -> recentPostDao.insertAll(recentPosts));
    }
}

File: app/src/main/java/im/point/dotty/repository/RecentRepo.kt
package im.point.dotty.repository

import android.annotation.SuppressLint
import im.point.dotty.db.RecentPostDao
import im.point.dotty.domain.AppState
import im.point.dotty.mapper.Mapper
import im.point.dotty.mapper.RecentPostMapper
import im.point.dotty.model.RecentPost
import im.point.dotty.network.MetaPost
import im.point.dotty.network.ObservableCallBackAdapter
import im.point.dotty.network.PointAPI
import im.point.dotty.network.PostsReply
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

class RecentRepo(private val api: PointAPI,
                 private val state: AppState,
                 private val recentPostDao: RecentPostDao,
                 private val mapper: Mapper<RecentPost, MetaPost> = RecentPostMapper()) : Repository<RecentPost> {

    @SuppressLint("CheckResult")
    private fun fetch(isBefore: Boolean): Single<List<RecentPost>> {
        val source = Observable.create { emitter: ObservableEmitter<PostsReply> ->
            api.getRecent(state.token  ?: throw Exception("invalid token"),
                    if (isBefore) state.recentPageId else null)
                    .enqueue(ObservableCallBackAdapter(emitter))}
                .observeOn(Schedulers.io())
                .flatMap { postsReply: PostsReply -> Observable.fromIterable(postsReply.posts) }
                .map { entry: MetaPost -> mapper.map(entry)}
        source.lastElement().subscribe { post -> state.recentPageId = post.pageId }
        return source.toList()
                .doOnSuccess { recentPosts: List<RecentPost> -> recentPostDao.insertAll(recentPosts)}
    }

    override fun getAll(): Flowable<List<RecentPost>> {
        return recentPostDao.getAll()
    }

    override fun fetch(): Single<List<RecentPost>> {
        return fetch(false)
    }

    fun fetchBefore(): Single<List<RecentPost>> {
        return fetch(true)
    }
}